File: leetcode/2588-CountTheNumberOfBeautifulSubarrays/solution.py
# Compute Complexity = O(N)
# Space Complexity = O(N)

class Solution:
    def beautifulSubarrays(self, nums: List[int]) -> int:
        xorHist = { 0: 1}
        sum = 0
        currXor = 0

        for n in nums:
            currXor ^= n
            if currXor in xorHist:
                sum += xorHist[currXor]
                xorHist[currXor] += 1
            else:
                xorHist[currXor] = 1

        return sum

File: leetcode/2588-CountTheNumberOfBeautifulSubarrays/solution.cpp
// Compute Complexity = O(N)
// Space Complexity = O(N)

class Solution {
public:
    long long beautifulSubarrays(vector<int>& nums) {
        std::unordered_map<int, long long> xorHist;
        xorHist[0] = 1;

        long long sum = 0;
        int currXor = 0;

        for (const auto& n: nums) {
            currXor ^= n;
            sum += xorHist[currXor];
            xorHist[currXor]++;
        }

        return sum;
    }
};